vmdb/app/models/metric/ci_mixin/capture/openstack_base.rb

```diff
@@ -37,134 +37,364 @@ def perf_release_openstack
+    multi_counter_aligned_end           = nil
+    # Guard for multicounter search distance for all counters data.
+    multi_counter_aligned_start_guard   = nil
+    # Bucket for collecting all related samples from multiple streams for certain period
+    multi_counter_metrics               = nil
+    # Bucket for collecting all related samples from multiple streams for certain period
+    last_multi_counter_metrics          = nil
+
+    timestamps.each_cons(2) do |last_period, period|
+      beginning_of_the_collection_period ||= last_period
+      multi_counter_aligned_start        ||= last_period
+      multi_counter_aligned_end          ||= period
+      multi_counter_aligned_start_guard  ||= last_period
+      multi_counter_aligned_start_guard    = period if multi_counter_aligned_start_guard == 'initialize_with_period'
+
+      multi_counter_metrics              ||= {}
+      last_multi_counter_metrics         ||= {}
+
+      metrics = {}
+      i[:openstack_counters].each { |c| metrics[c] = metrics_by_counter_name.fetch_path(c, period) }
+
+      # We need to make sure we first collect all data samples of last_metrics, then we can start to collect metrics
+      unless all_multi_counter_metrics_available?(i, last_multi_counter_metrics)
+        # All last multicounter data are not available, try to capture them
+        last_metrics = {}
+
+        if last_period - multi_counter_aligned_start_guard > data_collecting_period / 2
+          # If we haven't found all the data samples of all the counters half of the data_collecting_period away
+          # from the first data sample, just throw away everything and start over, cause we can't find it in this
+          # timeframe.
+          if beginning_of_the_collection_period == multi_counter_aligned_start
+            # If this is at the start of the whole collection period, it's not considered error. Due to
+            # overlapping of the collections periods, incomplete stat on the beginning had to be collected as part
+            # of the last collection period.
+            # That means we are moving the multi_counter_aligned_start so the incomplete period is entirely
+            # skipped.
+            multi_counter_aligned_start = last_period
           else
-            # the aggregate is still not divisible by 20, grab the next stat
-            next
+            # Not beginning, the data are corrupted or missing. In order to avoid holes in saved data, this period
+            # will be filled by data of the next periods.
+            $log.warn("#{log_header} id:[#{name}] Distance of the multiple streams of data is invalid. It "\
+                      "exceeded half of the Ceilometer collection period in "\
+                      "#{multi_counter_aligned_start_guard} - #{last_period} for counters" \
+                      "#{i[:openstack_counters]} It can be caused by different pipeline configuration "\
```

Comment:
- Need a period after the interpolation of `i[:openstack_counters]`.
- The second sentence needs grammar cleanup...perhaps: `It can be caused by a different pipeline configuration period for each related sample, or Ceilometer needs to scaled because the samples collection is overloaded.`